src/nifreeze/data/base.py

```diff
@@ -0,0 +1,283 @@
+    estimates).
+
+    The data structure has a direct HDF5 mapping to facilitate memory efficiency.
+    For modalities requiring additional metadata such as DWI (which requires the gradient table
+    and potentially a b=0 reference), this class may be derived to override certain behaviors
+    (in the case of DWIs, the indexed access should also return the corresponding gradient
+    specification).
+
+    """
+
+    dataobj = attr.ib(default=None, repr=_data_repr, eq=attr.cmp_using(eq=_cmp))
+    """A :obj:`~numpy.ndarray` object for the data array."""
+    affine = attr.ib(default=None, repr=_data_repr, eq=attr.cmp_using(eq=_cmp))
+    """Best affine for RAS-to-voxel conversion of coordinates (NIfTI header)."""
+    brainmask = attr.ib(default=None, repr=_data_repr, eq=attr.cmp_using(eq=_cmp))
+    """A boolean ndarray object containing a corresponding brainmask."""
+    motion_affines = attr.ib(default=None, eq=attr.cmp_using(eq=_cmp))
+    """List of :obj:`~nitransforms.linear.Affine` realigning the dataset."""
+    datahdr = attr.ib(default=None)
+    """A :obj:`~nibabel.spatialimages.SpatialHeader` header corresponding to the data."""
+
+    _filepath = attr.ib(
+        factory=lambda: Path(mkdtemp()) / "hmxfms_cache.h5",
+        repr=False,
+        eq=False,
+    )
+    """A path to an HDF5 file to store the whole dataset."""
+
+    def __len__(self) -> int:
+        """Obtain the number of volumes/frames in the dataset."""
+        if self.dataobj is None:
+            return 0
+
+        return self.dataobj.shape[-1]
+
+    def __getitem__(
+        self, idx: int | slice | tuple | np.ndarray
+    ) -> tuple[np.ndarray, np.ndarray | None]:
+        """
+        Returns volume(s) and corresponding affine(s) through fancy indexing.
+
+        Parameters
+        ----------
+        idx : :obj:`int` or :obj:`slice` or :obj:`tuple` or :obj:`~numpy.ndarray`
+            Indexer for the last dimension (or possibly other dimensions if extended).
+
+        Returns
+        -------
+        :obj:`tuple`
+            The selected volume(s) and corresponding affine(s).
+
+        """
+        if self.dataobj is None:
+            raise ValueError("No data available (dataobj is None).")
+
+        affine = self.motion_affines[idx] if self.motion_affines is not None else None
+        return self.dataobj[..., idx], affine
+
+    def get_filename(self) -> Path:
+        """Get the filepath of the HDF5 file."""
+        return self._filepath
+
+    def set_transform(self, index: int, affine: np.ndarray, order: int = 3) -> None:
+        """
+        Set an affine transform for a particular index and update the data object.
+
+        Parameters
+        ----------
+        index : :obj:`int`
+            The volume index to transform.
+        affine : :obj:`numpy.ndarray`
+            The 4x4 affine matrix to be applied.
+        order : :obj:`int`, optional
+            The order of the spline interpolation. Default is 3.
```

Comment:
I would remove redundant notes about default values in the docstrings: the method signature shows already what the default value is, and if for whatever reason the default value changes in the signature, it may happen that we miss to update the docstring.